eth/transaction_from_raw.go

```diff
@@ -1,178 +1,181 @@
-	// the R/S/V values.  Which for EIP155 transactions is the RLP encoding of the list:
-	//   [ Nonce, GasPrice, GasLimit, To, Value, Input, ChainId, 0, 0 ]
-	// While for non-EIP155 transactions (where ChainId is zero), only the first six fields are hashed:
-	//   [ Nonce, GasPrice, GasLimit, To, Value, Input ]
-	raw := rlp.Value{
-		List: []rlp.Value{
-			items[0], items[1], items[2], items[3], items[4], items[5],
-		},
-	}
+		t.Hash = raw.Hash()
+		t.From = *sender
+		return nil
+	default:
+		// Legacy Transaction
+		// Decode the input string as an rlp.Value
+		if err := rlpDecodeList(input, &nonce, &gasPrice, &gasLimit, &to, &value, &data, &v, &r, &s); err != nil {
+			return errors.Wrap(err, "could not decode RLP components")
+		}
 
-	// Compute the "recovery" V used when reversing the signing, which for non-EIP155 transactions
-	// is just V - 27.
-	recovery := v.Int64() - 27
+		// ... and fill in all our fields with the decoded values
+		t.Nonce = nonce
+		t.GasPrice = gasPrice
+		t.Gas = gasLimit
+		t.To = to
+		t.Value = value
+		t.Input = data
+		t.V = v
+		t.R = r
+		t.S = s
+
+		// Pull out chainId and recoveryV from EIP-155 packed V
+		_chain := (v.Int64() - 35) / 2
+		if _chain < 0 {
+			_chain = 0
+		}
 
-	if chainId != 0 {
-		// as mentioned above, append [ ChainId, 0, 0 ] to the hash input
-		raw.List = append(raw.List,
-			QuantityFromInt64(chainId).RLP(),
-			rlp.Value{String: "0x"},
-			rlp.Value{String: "0x"},
-		)
+		_recovery := v.Int64() - 27
+		if _chain != 0 {
+			// And subtract out the chainId to get a proper recovery value
+			_recovery -= (_chain * 2) + 8
+		}
 
-		// And subtract out the chainId to get a proper recovery value
-		recovery -= (chainId * 2) + 8
-	}
+		recoveryV := QuantityFromInt64(_recovery)
+		chainId = QuantityFromInt64(_chain)
 
-	recoveryV := QuantityFromInt64(recovery)
+		signingHash, err := t.SigningHash(chainId)
+		if err != nil {
+			return err
+		}
 
-	// We can now compute the input hash for address recovery
-	rh, err := raw.Hash()
-	if err != nil {
-		return errors.Wrap(err, "could not hash raw transaction")
-	}
-	recoverHash, err := NewHash(rh)
-	if err != nil {
-		return errors.Wrap(err, "invalid raw transaction hash")
-	}
+		sender, err := ECRecover(signingHash, &r, &s, &recoveryV)
+		if err != nil {
+			return err
+		}
 
-	// And now we can pass these values into ECRecover, which returns the sender address
-	sender, err := ECRecover(recoverHash, r, s, &recoveryV)
-	if err != nil {
-		return errors.Wrap(err, "could not recover from address")
-	}
+		raw, err := t.RawRepresentation(chainId)
+		if err != nil {
+			return err
+		}
 
-	// And finally, we need to compute the transaction's Hash...
-	dh, err := decoded.Hash()
-	if err != nil {
-		return errors.Wrap(err, "could not hash decoded transaction")
+		t.Hash = raw.Hash()
+		t.From = *sender
+		return nil
 	}
+}
 
-	txHash, err := NewHash(dh)
+func rlpDecodeList(input string, receivers ...interface{}) error {
```

Comment:
Perhaps a comment for this function discussing supported receivers would be useful.

Reply:
Added in ea668bd.